src/meander/interpreter/epsilon.cljc

```diff
@@ -0,0 +1,902 @@
+            [meander.util.epsilon :as m.util]
+            [meander.match.syntax.epsilon :as m.match.syntax]
+            [meander.syntax.epsilon :as m.syntax]))
+
+
+
+;; Runtime
+;; -------
+
+;; A runtime is a map of functions and values used to construct
+;; matching functions.
+;;
+;; Matching functions take an object called "target" and a map called
+;; "bindings" and return either a passing result with the runtime
+;; function "pass", or a failing result with the runtime value "fail".
+
+;; Results are process with the fuctions "fmap" which maps a function
+;; f over a result, and "join" which combines two results.
+
+;; Finally, there is the "scan" function which takes a sequence of
+;; objects and a function which takes any object and returns a result.
+
+(def default-eval
+  #?(:clj eval
+     :cljs (fn no-eval [_]
+             (throw (ex-info "eval not defined" {})))))
+
+(def
+  ^{:private true}
+  find-runtime
+  (let [fail (reify)]
+    {:eval default-eval
+     :pass identity
+     :fail fail
+     :fmap (fn [f x]
+             (if (identical? x fail)
+               x
+               (f x)))
+     :join (fn
+             ([a] a)
+             ([a b] a))
+     :scan (fn [p xs]
+             (reduce
+              (fn [_ x]
+                (let [y (p x)]
+                  (if (identical? y fail)
+                    y
+                    (reduced y))))
+              fail
+              xs))}))
+
+(def
+  ^{:private true}
+  depth-first-search-runtime
+  {:eval default-eval
+   :pass list
+   :fail ()
+   :fmap mapcat 
+   :join clojure/concat
+   :scan mapcat})
+
+(def
+  ^{:private true}
+  breadth-first-search-runtime
+  (let [fmap (fn g [f x]
+               (if (seq x)
+                 (m.util/mix (f (first x))
+                             (lazy-seq (g f (rest x))))
+                 ()))]
+    {:eval default-eval
+     :pass list
+     :fail ()
+     :fmap fmap
+     :join (fn
+             ([a] a)
+             ([a b] (m.util/mix a b)))
+     :scan fmap}))
+
+;; Matching function constructors
+;; ------------------------------
+
+;; Matching function constructors take a runtime along with other
+;; arguments and return a matching function. The prefix "p" or "p_" is
+;; used to indicate a matching function.
+
+(defn make-anything
+  {:private true}
+  [runtime]
+  (let [pass (get runtime :pass)]
+    (fn [target bindings]
+      (pass bindings))))
+
+(defn make-pred
+  {:private true}
+  [runtime p]
+  (let [pass (get runtime :pass)
+        fail (get runtime :fail)]
+    (fn [target bindings]
+      (if (p target)
+        (pass bindings)
+        fail))))
+
+(defn make-apply
+  {:private true}
+  [runtime f p]
+  (fn [target bindings]
+    (p (f target) bindings)))
+
+(defn make-all
+  {:private true}
+  [runtime p1 p2]
+  (let [fmap (get runtime :fmap)]
+    (fn [target bindings]
+      (fmap (fn [bindings] (p2 target bindings))
+            (p1 target bindings)))))
+
+
+(defn make-one
```

Comment:
@jimmyhmiller This isn't mapped to an AST node yet but this is the "committed choice" operator we discussed previously.